File: LeetCode/Explore/TopInterview/Amazon/ArrayString/48_rotate_image.cpp
class Solution {
  
private:
  void transpose(vector<vector<int> > &matrix) {
    for (int i = 0; i < matrix.size(); i++) {
      for (int j = 0; j < i; j++) {
        swap(matrix[i][j], matrix[j][i]);
      }
    }
  }

  void reversal(vector<vector<int> > &matrix) {
    for (int i = 0; i < matrix.size(); i++) {
      reverse(matrix[i].begin(), matrix[i].end());
    }
  }
  
public:
    void rotate(vector<vector<int>>& matrix) {
      transpose(matrix);
      reversal(matrix);
    }
};

File: LeetCode/Explore/TopInterview/Amazon/ArrayString/48_rotate_image.py
class Solution:
    def transpose(self, matrix: List[List[int]]):
      for i in range(len(matrix)):
        for j in range(i):
          tmp = matrix[i][j]
          matrix[i][j] = matrix[j][i]
          matrix[j][i] = tmp
          
    def reversal(self, matrix: List[List[int]]):
      for i in range(len(matrix)):
        matrix[i] = matrix[i][::-1]
    
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        self.transpose(matrix)
        self.reversal(matrix)
